src/System.ServiceModel.NetNamedPipe/src/System/ServiceModel/Channels/PipeConnection.cs

```diff
@@ -0,0 +1,576 @@
+                    try
+                    {
+                        await WaitForReadZero(readValueTask, timeout, true);
+                    }
+                    catch (TimeoutException e)
+                    {
+                        throw DiagnosticUtility.ExceptionUtility.ThrowHelper(
+                            new TimeoutException(SR.PipeShutdownReadError, e), ExceptionEventType);
+                    }
+                }
+                else if (existingReadIsPending)
+                {
+                    if (!TimeoutHelper.Wait(_atEOFEvent, timeoutHelper.RemainingTime()))
+                    {
+                        throw DiagnosticUtility.ExceptionUtility.ThrowHelper(
+                            new TimeoutException(SR.PipeShutdownReadError), ExceptionEventType);
+                    }
+                }
+                // else we had already seen EOF.
+
+                // at this point, we may get exceptions if the other side closes the handle first
+                try
+                {
+                    // write an ack for eof
+                    writeValueTask = StartWriteZeroAsync(cancellationToken);
+
+                    // read an ack for eof
+                    readValueTask = StartReadZeroAsync(cancellationToken);
+
+                    // wait for write to complete/fail
+                    await WaitForWriteZero(writeValueTask, timeout, false);
+
+                    // wait for read to complete/fail
+                    await WaitForReadZero(readValueTask, timeout, false);
+                }
+                catch (PipeException e)
+                {
+                    int errorCode = PipeError.GetErrorFromHResult(e.ErrorCode);
+                    if (!IsBrokenPipeError(errorCode))
+                    {
+                        throw;
+                    }
+                    DiagnosticUtility.TraceHandledException(e, TraceEventType.Information);
+                }
+                catch (CommunicationException e)
+                {
+                    DiagnosticUtility.TraceHandledException(e, TraceEventType.Information);
+                }
+                catch (TimeoutException e)
+                {
+                    DiagnosticUtility.TraceHandledException(e, TraceEventType.Information);
+                }
+            }
+            catch (TimeoutException e)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelper(
+                    new TimeoutException(SR.PipeCloseFailed, e), ExceptionEventType);
+            }
+            catch (PipeException e)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelper(
+                    ConvertPipeException(SR.PipeCloseFailed, e, TransferOperation.Undefined), ExceptionEventType);
+            }
+            finally
+            {
+                if (shouldCloseHandle)
+                {
+                    CloseHandle(false, null, TransferOperation.Undefined);
+                }
```

Comment:
nit: Does handle come from old code where native pipe types were used directly? I'd change handle to pipe in names.